src/shogun/io/SGIO.h

```diff
@@ -84,124 +87,72 @@ enum EMessageLocation
+		    MSG_ERROR, _SRC_LOC, __VA_ARGS__);  \
 	}
 #define SG_ERROR(...) SG_THROW(ShogunException, __VA_ARGS__)
-#define SG_OBJ_ERROR(o, ...)                                                   \
-	{                                                                          \
-		o->io->template error<ShogunException>(                                \
-		    MSG_ERROR, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);  \
-	}
-#define SG_CLASS_ERROR(c, ...)                                                 \
-	{                                                                          \
-		c::io->template error<ShogunException>(                                \
-		    MSG_ERROR, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);  \
-	}
-#define SG_UNSTABLE(func, ...) { io->message(MSG_WARN, __PRETTY_FUNCTION__, __FILE__, __LINE__, \
+
+#define SG_UNSTABLE(func, ...) { SG_IO->message(MSG_WARN, _SRC_LOC, \
 __FILE__ ":" func ": Unstable method!  Please report if it seems to " \
 "work or not to the Shogun mailing list.  Thanking you in " \
 "anticipation.  " __VA_ARGS__); }
 
-#define SG_PRINT(...) { io->message(MSG_MESSAGEONLY, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__); }
-#define SG_OBJ_PRINT(o, ...) { o->io->message(MSG_MESSAGEONLY, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__); }
-#define SG_NOTIMPLEMENTED { io->not_implemented(__PRETTY_FUNCTION__, __FILE__, __LINE__); }
-#define SG_GPL_ONLY { io->gpl_only(__PRETTY_FUNCTION__, __FILE__, __LINE__); }
+#define SG_PRINT(...) { SG_IO->message(MSG_MESSAGEONLY, _SRC_LOC, __VA_ARGS__); }
+#define SG_NOTIMPLEMENTED { SG_IO->not_implemented(_SRC_LOC); }
+#define SG_GPL_ONLY { SG_IO->gpl_only(_SRC_LOC); }
 
 #define SG_DONE() {								\
-	if (SG_UNLIKELY(io->get_show_progress()))	\
-		io->done();								\
-}
-
-// printf like function using the global SG_IO object
-#define SG_IO env()->io()
-
-#define SG_SGCDEBUG(...) {											\
-	if (SG_UNLIKELY(SG_IO->loglevel_above(MSG_GCDEBUG)))			\
-		SG_IO->message(MSG_GCDEBUG,__PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);\
-}
-
-#define SG_SDEBUG(...) {											\
-	if (SG_UNLIKELY(SG_IO->loglevel_above(MSG_DEBUG)))			\
-		SG_IO->message(MSG_DEBUG,__PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);	\
-}
-
-#define SG_SINFO(...) {												\
-	if (SG_UNLIKELY(SG_IO->loglevel_above(MSG_INFO)))				\
-		SG_IO->message(MSG_INFO,__PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);	\
-}
-
-#define SG_SWARNING(...) { SG_IO->message(MSG_WARN,__PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__); }
-#define SG_STHROW(Exception, ...)                                              \
-	{                                                                          \
-		SG_IO->template error<Exception>(                                      \
-		    MSG_ERROR, __PRETTY_FUNCTION__, __FILE__, __LINE__, __VA_ARGS__);  \
-	}
-#define SG_SERROR(...) SG_STHROW(ShogunException, __VA_ARGS__)
```

Comment:
no need for the "SG_S{logleve}" class of macros

Reply:
Can you briefly point me to why this is not needed anymore?

Reply:
SGObject's member variable "io" is removed, and only the global object is used "env()->io().